tabpopup: Draw popups as OSD windows

vkareh opened the pull request:
This changes how the Alt+Tab and Workspace Switcher popups are drawn. The are now styled as OSD windows with a bunch of stylistic improvements. Also more things are drawn as cairo surfaces and some gschema values have been changed to make them compatible with the new design.

vkareh commented:
Alt+Tab Switcher (icons, composited):

![icons](https://user-images.githubusercontent.com/259780/66780172-2be23800-ee9e-11e9-9d49-31ef9c6a5bf7.png)

Alt+Tab Switcher (thumbnails):

![thumbnails](https://user-images.githubusercontent.com/259780/66780174-2be23800-ee9e-11e9-86f0-2ec93ed858dc.png)

Alt+Tab Switcher (non-composited):

![non-composited](https://user-images.githubusercontent.com/259780/66780173-2be23800-ee9e-11e9-9f43-cc98a44e81d9.png)

Workspace Switcher:

![workspaces](https://user-images.githubusercontent.com/259780/66780176-2be23800-ee9e-11e9-960d-3ac4611779aa.png)

rbuj commented on src/include/common.h:
```diff
@@ -319,9 +319,9 @@ struct _MetaFrameBorders
-#define META_DEFAULT_ICON_SIZE 48
-#define META_MIN_ICON_SIZE 6
-#define META_MAX_ICON_SIZE 480
+#define META_DEFAULT_ICON_SIZE 64
```
It's better to keep `#define META_DEFAULT_ICON_SIZE 48` for non HiDPI resolutions.

vkareh replied:
What resolution do you use? I tested on both HiDPI and regular, but now that I think of it I only tried one resolution (1920x1080, or 3840x2160 for HiDPI). I'll test all possible resolutions tomorrow to make sure things work

rbuj replied:
I used 2560x1600 for HiDPI and on 1920x1200 for regular.

rbuj commented on src/org.mate.marco.gschema.xml:
```diff
@@ -197,13 +197,13 @@
-        <range min="6" max="480"/>
-        <default>48</default>
+        <range min="8" max="256"/>
+        <default>64</default>
```
It's better to keep `<default>48</default>` for non HiDPI resolutions.

sc0w commented:
New warnings detected in this PR:

```
ui/tabpopup.c: In function 'popup_window_screen_changed':
ui/tabpopup.c:118:41: warning: unused parameter 'old_screen' [-Wunused-parameter]
  118 |                              GdkScreen *old_screen,
      |                              ~~~~~~~~~~~^~~~~~~~~~

ui/tabpopup.c:119:41: warning: unused parameter 'data' [-Wunused-parameter]
  119 |                              gpointer   data)
      |                              ~~~~~~~~~~~^~~~

ui/tabpopup.c: In function 'meta_select_image_draw':
ui/tabpopup.c:852:24: warning: declaration of 'context' shadows a previous local [-Wshadow]
  852 |       GtkStyleContext *context;
      |                        ^~~~~~~

ui/tabpopup.c:843:20: note: shadowed declaration is here
  843 |   GtkStyleContext *context;
      |                    ^~~~~~~

ui/tabpopup.c:877:7: warning: implicit declaration of function 'meta_gtk_style_get_light_color' [-Wimplicit-function-declaration]
  877 |       meta_gtk_style_get_light_color (context, GTK_STATE_FLAG_SELECTED, &color);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

ui/tabpopup.c:877:7: warning: nested extern declaration of 'meta_gtk_style_get_light_color' [-Wnested-externs]

src/core/screen.c:1311:11: style: The scope of the variable 'width' can be reduced. [variableScope]
      int width = 0, height = 0;
          ^

src/core/screen.c:1311:22: style: The scope of the variable 'height' can be reduced. [variableScope]
      int width = 0, height = 0;
                     ^

src/core/screen.c:1326:38: style: The scope of the variable 'icon' can be reduced. [variableScope]
          cairo_surface_t *surface, *icon, *win_surface;
                                     ^

src/core/screen.c:1327:20: style: The scope of the variable 'cr' can be reduced. [variableScope]
          cairo_t *cr;
                   ^

src/core/screen.c:1328:15: style: The scope of the variable 'icon_width' can be reduced. [variableScope]
          int icon_width, icon_height;
              ^

src/core/screen.c:1328:27: style: The scope of the variable 'icon_height' can be reduced. [variableScope]
          int icon_width, icon_height;
                          ^

src/ui/tabpopup.c:852:24: style: Local variable 'context' shadows outer variable [shadowVariable]
      GtkStyleContext *context;
                       ^

src/ui/tabpopup.c:843:20: note: Shadowed declaration
  GtkStyleContext *context;
                   ^

src/ui/tabpopup.c:852:24: note: Shadow variable
      GtkStyleContext *context;
                       ^
```

vkareh commented:
@sc0w - thanks for the warnings, I've fixed most of them (except the `*old_screen` and `data`, as those are from a `g_signal_connect`)
@rbuj - I'll test some more tomorrow, as I'm not getting the same sizes as you for some reason...

sc0w commented:
@vkareh 

```patch
diff --git a/src/ui/tabpopup.c b/src/ui/tabpopup.c
index 26dcbb9..263a6a9 100644
--- a/src/ui/tabpopup.c
+++ b/src/ui/tabpopup.c
@@ -116,8 +116,8 @@ outline_window_draw (GtkWidget *widget,
 
 static void
 popup_window_screen_changed (GtkWidget *widget,
-                             GdkScreen *old_screen,
-                             gpointer   data)
+                             GdkScreen *old_screen G_GNUC_UNUSED,
+                             gpointer   data G_GNUC_UNUSED)
 {
   /* To check if the display supports alpha channels, get the visual */
   GdkScreen *screen = gtk_widget_get_screen (widget);

vkareh commented:
@sc0w - thanks, `G_GNUC_UNUSED` is exactly what I needed :+1:

@rbuj - I've updated the default icon size to `48`

raveit65 commented:
off topic on
Any chance to fix GtkMisc deprecation when we already do changes in tabpopup.c ?
GtkMisc is one of the first deprecations started with Gtk+-3.0 and it is already fixed in metacity.
I guess with that PR we lost any chances to adopt this code.
Thanks.
offtopic off

raveit65 commented:
Not sure if this is related to PR but description of gsettings key icon-size can be improved.
Nobody knows for what is good for before you double-click selected key in dconf-editor.
![Bildschirmfoto zu 2019-10-21 00-41-46](https://user-images.githubusercontent.com/961604/67167579-fb1d6980-f39b-11e9-9a93-f2065d162428.png)

raveit65 commented:
I think this gsettings value should be the default with the nice new Look&feel , or not?
```
[rave@mother ~]$ gsettings get org.mate.Marco.general compositing-fast-alt-tab
false
```
Default is `true` which disables thumbnails.
The key and the negotiation in key description is really confusing.

vkareh commented:
@mate-desktop/core-team - can we vote on this gsettings change? Do we default to thumbnails (current default) or icons (proposed default)?

lukefromdc commented:
Alt-tab thumbnail speed could vary with many things. I have AMD bulldozer and a small first-gen GCN AMD graphics card and there is a real hesitation in startup with thumbnails. Thus I vote for icons as the default and a clear description of what this gsettings key is for.  Would be nice (but another PR to another package) to add a GUI function for toggling that in mate-control-center->windows

vkareh commented:
@raveit65 - I added a commit to fix all deprecations in tabpopup.c. I also fixed the text under the `icon-size` gsettings entry.

Since there doesn't seem to be a lot of consensus, I've decided to revert the default value of `compositing-fast-alt-tab` back to `false`. This way the current behavior is preserved and nothing changes in that respect.

@lukefromdc - I've started working on adding some of these settings to `mate-window-properties`, and I might change some marco gsettings as part of that, which would require a new version dependency on m-c-c. So I might still change the default value of `compositing-fast-alt-tab`, but if I do it would be as part of a different patch set.

vkareh commented:
Ready to merge for me! :) thanks for reviewing!

On October 27, 2019 9:19:11 AM EDT, raveit65 <notifications@github.com> wrote:
>raveit65 approved this pull request.
>
>This is a very nice and modern design  for the popups and it works
>well.
>Description of icon-size of tab-popup and deprecations are fixed now.
>Now marco is nearly deprecation free.
>You did really a good job.
>Thank you.
>
>PS: Is it ready merge?
>
>
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/mate-desktop/marco/pull/566#pullrequestreview-307572283
